src/main/java/com/scalar/cassy/transferer/AzureBlobFileDownloader.java

```diff
@@ -1,63 +1,103 @@
 package com.scalar.cassy.transferer;
 
-import com.azure.storage.blob.BlobContainerAsyncClient;
+import com.azure.storage.blob.BlobContainerClient;
 import com.azure.storage.blob.models.BlobItem;
-import com.azure.storage.blob.models.BlobProperties;
 import com.azure.storage.blob.models.ListBlobsOptions;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.inject.Inject;
 import com.scalar.cassy.config.RestoreConfig;
 import com.scalar.cassy.exception.FileTransferException;
+import java.io.FileOutputStream;
 import java.io.IOException;
+import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import reactor.core.publisher.Mono;
 
 public class AzureBlobFileDownloader implements FileDownloader {
   private static final Logger logger = LoggerFactory.getLogger(AzureBlobFileDownloader.class);
-  private BlobContainerAsyncClient blobContainerClient;
+  private BlobContainerClient blobContainerClient;
+  private static final int NUM_THREADS = 3;
+  private static final int ASYNC_FILE_DOWNLOAD_LIMIT = 20;
+  private final ExecutorService executorService = Executors.newFixedThreadPool(NUM_THREADS);
 
   @Inject
-  public AzureBlobFileDownloader(BlobContainerAsyncClient blobContainerClient) {
+  public AzureBlobFileDownloader(BlobContainerClient blobContainerClient) {
     this.blobContainerClient = blobContainerClient;
   }
 
   @Override
   public void download(RestoreConfig config) {
+    List<Future> downloadFuture = new ArrayList<>();
     String key = BackupPath.create(config, config.getKeyspace());
 
     logger.info("Downloading " + blobContainerClient.getBlobContainerName() + "/" + key);
-    List<Mono<BlobProperties>> filesToBeDownloaded = new ArrayList<>();
-    Iterable<BlobItem> keyspaceBlobs =
-        blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/")).toIterable();
+    Iterable<BlobItem> keyspaceBlobs = listBlobs(key);
     for (BlobItem blob : keyspaceBlobs) {
```

Comment:
```suggestion
    for (BlobItem blob : listBlobs(key)) {
```

Reply:
Thank you for your suggestion.
Suggestion updated cdb813996195945174f3e7614c63035b17e6e91e

> I don't get which part you feel difficult.
Can you clarify?

By default, we can simply create a blob file list
```
Iterable<BlobItem> keyspaceBlobs = blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/"), null);
```

In the above case, there is some difficulty for mock `PagedIterable<BlobItem>` so we changed the source code with an additional method like this for the test case
```
  @VisibleForTesting
  Iterable<BlobItem> listBlobs(String key) {
    return blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/"), null);
  }
```

If we use `iterator`, we can avoid the additional method ` listBlobs(String key)` from the source code.

That's what I meant

Reply:
https://github.com/scalar-labs/cassy/pull/76#discussion_r647057830

Reply:
https://github.com/scalar-labs/cassy/pull/76#discussion_r647066442